### ISSUE TITLE
Dual monitor configuration - issues overview mode

### ISSUE THREAD

**Issue by bauno** [micheleg/dash-to-dock#410]
Dual monitor configuration, primary on the front, secondary on the left.
I keep the dash to dock extension on the right of the secondary monitor.
When I switch in overview mode (SUPER-S) the workspaces on the right on the primary monitor is shifted on the left of the edge of the screen...seems to leave space for the dash (which is on the other monitor).
If I change the position of the dash, everything is ok.
A screenshot explains this better than I do.

Fantastic extension BTW...thanks!
A.
![screenshot from 2016-11-28 12-51-33](https://cloud.githubusercontent.com/assets/1089931/20667666/af6bef2e-b56a-11e6-8a18-c740a73b7f7f.png)

**Comment by micheleg** [comment]
Thanks for your bug report. I'll try to debug this. it definitely looks like the space for the dock is reserved, although it should not be applied with the dock on the secondary monitor. I see that you have somehow tweaked to position of the workspace thumbnails too. Could you double check if the problem is connected with an incompatibility with some other extension? 

Also, please report your extension and gnome version.

**Comment by bauno** [comment]
Hi Michele,

Thanks for your reply.
I disabled all the other extensions but the problem persists.
I use Gnome 3.22 on Fedora 25, started using on Fedora 24/Gnome 3.20 (same problem from "a moment" onward)
Dash-to-Dock version is 55.

Thanks!
A.

**Comment by franglais125** [comment]
@bauno I'm trying to reproduce your issue, but I can't at the moment. I think the reason is that you have somehow **two** workspace thumnails. How did you tweak this? (I think this is what @micheleg was asking). Here is a screenshot of my screen, with only one set of workspace thumnbails (will upload shortly):

**Comment by bauno** [comment]
Ahhh now I get it. I use the "Mujltiple monitor extension": https://goo.gl/vyjx0r
I disabled it, but no go...:-?

**Comment by franglais125** [comment]
@bauno I managed to reproduce your problem, without any other extension enabled.

@micheleg I think I have figured out what the problem is... In docking.js we are creating a dashSpacer and assigning it to the main monitor in every case (actually, wherever the overview is), even if the dock is on a secondary monitor. I only now found this (but not critical I would say).

When the dock is placed on a secondary monitor:
 - There is an extra spacing on the main monitor (this bug)
 - No spacing on a secondary monitor (see screenshot below, to be uploaded soon)

The multi-monitor add-on @bauno mentioned essentially creates a new overview for secondary monitors. It looks like something like this would need to be done to correctly assign the dashSpacer when the dock is on a secondary monitor... All that sounds like a pain. Any ideas? For the time being, we could at least not assign any spacing when the dock is on a secondary monitor.

**Comment by bauno** [comment]
Thanks guys...

**Comment by micheleg** [comment]
I thing the spacer should not have any effect if the dock is not in the primary monitor given:

https://github.com/micheleg/dash-to-dock/blob/master/docking.js#L985-L991

So why is is happening?

**Comment by franglais125** [comment]
@micheleg I think it's in the other extension's code:
https://github.com/spin83/multi-monitors-add-on/blob/master/multi-monitors-add-on%40spin83/mmoverview.js#L491

They are taking the geometry from the primary monitor it seems?

**Comment by micheleg** [comment]
Sorry for the boring question, but I don't have a second monitor to test now. Does any problem occurs without the "Multiple monitor extension"?

**Comment by franglais125** [comment]
With the other extension **disabled**, the only problem I see is that windows overlap [1] with the dash in the overview, but this is due to the spacer being hidden. This is with the dash on a secondary monitor.

[1] https://cloud.githubusercontent.com/assets/19195975/21026895/9dbd4a3a-bd5c-11e6-8454-410a4b8700a0.png

What @bauno experiences with the other extension **enabled** is an extra spacing, but I think it is the other extension which is assigning a new spacer.

**Comment by micheleg** [comment]
Ok, so on out side ideally we should have a spacer in the secondary monitor too, while I think the issue with the other extension is in the way they build the additional overview structure.

**Comment by bauno** [comment]
Sorry guys but it seems to happen even with MM disabled...
![screenshot from 2016-12-11 18-22-59](https://cloud.githubusercontent.com/assets/1089931/21081835/1d5037fc-bfcf-11e6-80f2-e5df9e31797a.png)

**Comment by franglais125** [comment]
@bauno Indeed, but I still think this is actually equivalent to the one I just posted
Here, I upload a new screenshot that illustrates your case and my case at the same time (see below).

In both cases, the spacer is shown in the primary monitor, regardless of where the dash actually is.
Hence, when you place the dash on the right of the secondary monitor, you have:
- an extra space in the primary monitor (your screenshot)
- a missing space in the secondary monitor (the overlap I was showing)

Now, to see both at the same time:

![screenshot from 2016-12-11 12 36 49](https://cloud.githubusercontent.com/assets/19195975/21081930/b05f0bf8-bf9e-11e6-91cc-1eeaffd497b5.png)

**Comment by franglais125** [comment]
(Re-posting from another issue) @micheleg I've seen this issue many times when working on the multi-monitor code. The problem is the `dashSpacer` gets created for every dock instance, but the workspace code upstream doesn't use its geometry to properly calculate the space for the window spread except if in the primary screen). Here is the relevant code upstream: https://github.com/GNOME/gnome-shell/blob/033277b68f2910e54ce62f45fa92af2f5092a7ba/js/ui/workspacesView.js#L658

I tried to tackle this, but honestly just ran out of time. I'm leaving my work in this branch in case somebody else wants to try and fix this: https://github.com/franglais125/dash-to-dock/tree/multimon_dash_spacer

The best part is that I managed to identify the part of the code that needs some tweaking. It "works", as in it almost solves the issue, but needs some polishing:

- I basically just forced a new injection, maybe there is a better way of doing it.
- For some reason, after injection, the fix does not apply immediately. It always takes me 3 opening/closing the overview before the geometry is recalculated.
- Code style and structure to be improved for sure.

Cheers

**Comment by franglais125** [comment]
I just updated the branch with the fix (https://github.com/franglais125/dash-to-dock/tree/multimon_dash_spacer), I addresses the 3 issues I listed above. (Thanks @passingthru67 for the help!)

I now just need to fix the last glitch: after screen lock/unlock, for some reason the geometry doesn't update properly anymore, probably something to do with injection. I'll take a look later.

**Comment by franglais125** [comment]
Ok, finally, the fix seems to be working fine! Feedback welcome.

**Comment by micheleg** [comment]
@franglais125 Thanks for working on this. I'm also lacking time lately (and as usual I cannot easily provide feedback for multimonitor... even my secondary monitor at work died). The solution seems reasonable to me.

**Comment by franglais125** [comment]
> The solution seems reasonable to me.

Great, hopefully we can get feedback from someone else confirming the fix. @bauno do you mind testing this branch https://github.com/franglais125/dash-to-dock/tree/multimon_dash_spacer ?

------------------------
Just for completeness: regarding the code, the injection of the dummy function `setActualGeometry = {return;}` is due to the upstream signal

```
this.actor.connect('notify::allocation', Lang.bind(this, this._updateWorkspacesActualGeometry));
```

that doesn't save the `id`, so we can't remove it. If this signal id were saved in the object, we could reconnect it properly.

**Comment by bauno** [comment]
First of all, guys, thanks for all the work.
Second: the new extension fails to start with the following error:

**Comment by franglais125** [comment]
Yeah, you really need to `make install` for this to work properly.

```
git clone https://github.com/franglais125/dash-to-dock/
cd dash-to-dock
git checkout multimon_dash_spacer
make install
```

Then reload the Shell (or log out if in Wayland).

**Comment by bauno** [comment]
@franglais125, sorry but I didn't see your last reply and didn't try installing the updated extension.
I did it now, and the extension loads correctly, but nothing has changed.
Sorry again for the very late answer.

**Comment by vanvugt** [comment]
This appears to be [gnome-shell#517](https://gitlab.gnome.org/GNOME/gnome-shell/-/issues/517) so can probably be closed.
